internal/luks2/metadata.go

```diff
@@ -0,0 +1,910 @@
+	Segments   []int   // The segments assigned to this digest
+	Salt       []byte  // Salt for this digest
+	Digest     []byte  // The actual digest
+	Hash       Hash    // Hash algorithm (pbkdf2 only)
+	Iterations int     // The number of iterations (pbkdf2 only)
+}
+
+func (d *Digest) UnmarshalJSON(data []byte) error {
+	var t struct {
+		Type       KDFType
+		Keyslots   []luksJsonNumber
+		Segments   []luksJsonNumber
+		Salt       []byte
+		Digest     []byte
+		Hash       Hash
+		Iterations int
+	}
+	if err := json.Unmarshal(data, &t); err != nil {
+		return err
+	}
+
+	*d = Digest{
+		Type:       t.Type,
+		Salt:       t.Salt,
+		Digest:     t.Digest,
+		Hash:       t.Hash,
+		Iterations: t.Iterations}
+
+	for _, v := range t.Keyslots {
+		s, err := v.int()
+		if err != nil {
+			return xerrors.Errorf("invalid keyslot id: %w", err)
+		}
+		d.Keyslots = append(d.Keyslots, s)
+	}
+
+	for _, v := range t.Segments {
+		s, err := v.int()
+		if err != nil {
+			return xerrors.Errorf("invalid segment id: %w", err)
+		}
+		d.Segments = append(d.Segments, s)
+	}
+
+	return nil
+}
+
+// Integrity corresponds to an integrity object in the JSON metadata of a LUKS2 volume,
+// and details the data integrity parameters for a segment.
+type Integrity struct {
+	Type              string // Integirty type in dm-crypt notation
+	JournalEncryption string
+	JournalIntegrity  string
+}
+
+// Segment corresponds to a segment object in the JSON metadata of a LUKS2 volume,
+// and details an encrypted area on disk.
+type Segment struct {
+	Type        string
+	Offset      uint64     // Offset from the device start to the beginning of this segment, in bytes
+	Size        uint64     // Size of this segment, in bytes (only if DynamicSize is false)
+	DynamicSize bool       // The size is the size of the underlying device
+	IVTweak     uint64     // The starting offset of the IV tweak
+	Encryption  string     // The encryption algorithm for this segment in dm-crypt notation
+	SectorSize  int        // The sector size for this segment, in bytes
+	Integrity   *Integrity // Data integrity parameters for this segment (optional)
+	Flags       []string   // Additional options for this segment
+}
+
+func (s *Segment) UnmarshalJSON(data []byte) error {
+	var d struct {
+		Type       string
+		Offset     luksJsonNumber
+		Size       luksJsonNumber
+		IVTweak    luksJsonNumber `json:"iv_tweak"`
+		Encryption string
+		SectorSize int `json:"sector_size"`
+		Integrity  *Integrity
+		Flags      []string
+	}
+	if err := json.Unmarshal(data, &d); err != nil {
+		return err
+	}
+
+	*s = Segment{
+		Type:       d.Type,
+		Encryption: d.Encryption,
+		SectorSize: d.SectorSize,
+		Integrity:  d.Integrity,
+		Flags:      d.Flags}
+
+	offset, err := d.Offset.uint64()
+	if err != nil {
+		return xerrors.Errorf("invalid offset value: %w", err)
+	}
+	s.Offset = offset
+
+	switch string(d.Size) {
+	case "dynamic":
+		s.DynamicSize = true
+	default:
+		sz, err := d.Size.uint64()
```

Comment:
this case doesn't seem to be exercised

Reply:
@chrisccoulson this is stil the case it seems? I know you said it was a bit complicated, what's the status on that?